src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+#include "storage/ipc.h"
+#include "storage/procarray.h"
+#include "storage/shm_mq.h"
+#include "storage/shm_toc.h"
+#include "tcop/pquery.h"
+#include "tcop/tcopprot.h"
+#include "tcop/utility.h"
+#include "utils/memutils.h"
+#include "utils/portal.h"
+#include "utils/ps_status.h"
+#include "utils/resowner.h"
+#include "utils/snapmgr.h"
+#include "utils/timeout.h"
+
+#include "distributed/background_jobs.h"
+#include "distributed/citus_safe_lib.h"
+#include "distributed/listutils.h"
+#include "distributed/maintenanced.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/shard_cleaner.h"
+#include "distributed/resource_lock.h"
+
+bool BackgroundTaskMonitorDebugDelay = false;
+
+/* Table-of-contents constants for our dynamic shared memory segment. */
+#define CITUS_BACKGROUND_TASK_MAGIC 0x51028081
+#define CITUS_BACKGROUND_TASK_KEY_DATABASE 0
+#define CITUS_BACKGROUND_TASK_KEY_USERNAME 1
+#define CITUS_BACKGROUND_TASK_KEY_COMMAND 2
+#define CITUS_BACKGROUND_TASK_KEY_QUEUE 3
+#define CITUS_BACKGROUND_TASK_KEY_TASK_ID 4
+#define CITUS_BACKGROUND_TASK_NKEYS 5
+
+static BackgroundWorkerHandle * StartCitusBackgroundJobExecuter(char *database,
+																char *user, char *command,
+																int64 taskId,
+																dsm_segment **pSegment);
+static void ExecuteSqlString(const char *sql);
+static void ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task,
+									bool *hadError, bool *cancelled);
+static void UpdateDependingTasks(BackgroundTask *task);
+
+PG_FUNCTION_INFO_V1(citus_job_cancel);
+PG_FUNCTION_INFO_V1(citus_job_wait);
+
+
+/*
+ * pg_catalog.citus_job_cancel(jobid bigint) void
+ *   cancels a scheduled/running job
+ *
+ * When cancelling a job there are two phases.
+ *  1. scan all associated tasks and transition all tasks that are not already in their
+ *     terminal state to cancelled. Except if the task is currently running.
+ *  2. for all running tasks we send a cancelation signal to the backend running the
+ *     query. The background executor/monitor will transition this task to cancelled.
+ *
+ * We apply the same policy checks as pg_cancel_backend to check if a user can cancel a
+ * job.
+ */
+Datum
+citus_job_cancel(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	int64 jobid = PG_GETARG_INT64(0);
+
+	/* Cancel all tasks that were scheduled before */
+	List *pids = CancelTasksForJob(jobid);
+
+	/* send cancellation to any running backends */
+	int pid = 0;
+	const int sig = SIGINT;
+	foreach_int(pid, pids)
+	{
+		PGPROC *proc = BackendPidGetProc(pid);
+		if (proc == NULL)
+		{
+			/*
+			 * This is just a warning so a loop-through-resultset will not abort
+			 * if one backend terminated on its own during the run.
+			 */
+			ereport(WARNING,
+					(errmsg("PID %d is not a PostgreSQL backend process", pid)));
+
+			continue;
+		}
+
+#ifdef HAVE_SETSID
+		if (kill(-pid, sig))
+#else
+		if (kill(pid, sig))
```

Comment:
I was fearing that also updating the status here would open the door for 2 processes to concurrently try to update the Task. Given the monitor has read the , possibly quite a while ago, into memory updating it from here would cause quite some headaches.